File: demo/gen_voiceover.py
# Generates the Trace demo voiceover pack (male Indian English narrator) via ElevenLabs.
# Each segment maps 1:1 to a beat in DEMO_SCRIPT.md.
import os, sys, json, urllib.request

# Read the key from the environment so it never lands in git:
#   ELEVENLABS_API_KEY=sk_... python gen_voiceover.py
API_KEY = os.environ.get("ELEVENLABS_API_KEY", "")
if not API_KEY:
    sys.exit("Set ELEVENLABS_API_KEY in your environment first (the audio is already generated in voiceover/).")
VOICE_ID = "fnyqY1A0BcjAtRpVEd7Z"          # Satyam Singh - Indian Business Narrator (confident)
MODEL = "eleven_multilingual_v2"            # highest-quality (non-realtime) narration
OUT = os.path.join(os.path.dirname(__file__), "voiceover")

# (id, filename, text). Confident narrator ~2.3 wps. Tightened to ~100s total so the
# finished cut (with on-screen action) stays comfortably under 2:00.
SEGMENTS = [
    ("01_cold_open", "01_cold_open.mp3",
     "Every A.I. call is stateless — it forgets. Your team decides something on Monday; "
     "by Friday the A.I. has no memory of it. "
     "This is Trace: team memory that never forgets, built on Cognee."),

    ("02_discord_drift", "02_discord_drift.mp3",
     "Live in Discord: a teammate says, let's move the new service to MongoDB. "
     "But this team standardized on Postgres months ago. "
     "Trace catches it instantly — citing the exact decision, and who made it. "
     "That's Cognee's graph memory reasoning over the team's history."),

    ("03_slack_duplicate", "03_slack_duplicate.mp3",
     "Same memory, in Slack. Someone starts a retry queue two teammates already built. "
     "Trace flags the duplicate before an hour is wasted. One brain, every channel."),

    ("04_github_pr", "04_github_pr.mp3",
     "The moneyshot: a pull request migrates billing to MongoDB — reversing a chat decision from weeks ago. "
     "Trace comments on the pull request itself, citing the original call. "
     "A decision made in Discord, now enforced on GitHub."),

    ("05_lifecycle", "05_lifecycle.mp3",
     "Under the hood: the full Cognee lifecycle. Every message is remembered — "
     "added and cognified into a hybrid graph-vector store. "
     "Ask anything; it recalls with citations, routing vector search and graph traversal automatically."),

    ("06_whatif", "06_whatif.mp3",
     "It even projects the future. If Priya leaves, what breaks? Authentication — she owns it, "
     "and she's on leave. Every consequence traced to a real message. Zero invented numbers."),

    ("07_close", "07_close.mp3",
     "Trace runs on Cognee Cloud and self-hosted — remember, recall, improve, forget. "
     "Give your A.I. a memory that survives the night. No more hangover."),
]

def tts(text, path):
    body = json.dumps({
        "text": text,
        "model_id": MODEL,
        "voice_settings": {"stability": 0.45, "similarity_boost": 0.8, "style": 0.35, "use_speaker_boost": True},
    }).encode()
    req = urllib.request.Request(
        f"https://api.elevenlabs.io/v1/text-to-speech/{VOICE_ID}?output_format=mp3_44100_128",
        data=body, method="POST",
        headers={"xi-api-key": API_KEY, "Content-Type": "application/json", "Accept": "audio/mpeg"},
    )
    with urllib.request.urlopen(req, timeout=120) as r:
        data = r.read()
    with open(path, "wb") as f:
        f.write(data)
    return len(data)

if __name__ == "__main__":
    os.makedirs(OUT, exist_ok=True)
    total = 0
    for sid, fname, text in SEGMENTS:
        path = os.path.join(OUT, fname)
        try:
            n = tts(text, path)
            total += n
            print(f"[ok] {fname:26s} {n//1024:5d} KB  ({len(text.split())} words)")
        except Exception as e:
            print(f"[FAIL] {fname}: {e}")
            sys.exit(1)
    print(f"Done. {len(SEGMENTS)} segments, {total//1024} KB total -> {OUT}")

File: scripts/gen-voiceover.mjs
// Generate the Trace demo voiceover with the ElevenLabs API.
// Narrator = one expressive voice; lines are placed per beat in the Remotion timeline.
// Usage:  ELEVENLABS_API_KEY=sk_... node scripts/gen-voiceover.mjs
// Output: demo/video/public/audio/vo/n01.mp3 ...  + a manifest.json with durations.
//
// v3 audio tags ([excited] etc.) are kept in the script for reference but STRIPPED
// before the API call and expressed via per-line voice_settings, so this works on any
// key/model. Set MODEL=eleven_v3 if your key has v3 access to keep the tags instead.

import { mkdirSync, writeFileSync, existsSync, statSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const here = dirname(fileURLToPath(import.meta.url));
const OUT = join(here, "..", "demo", "video", "public", "audio", "vo");
mkdirSync(OUT, { recursive: true });

const KEY = process.env.ELEVENLABS_API_KEY || process.env.ELEVEN_API_KEY;
if (!KEY) throw new Error("Set ELEVENLABS_API_KEY");
const MODEL = process.env.MODEL || "eleven_multilingual_v2"; // set eleven_v3 if you have access
const NARRATOR = process.env.NARRATOR_VOICE || "IKne3meq"; // Charlie — Deep, Confident, Energetic

// mood → voice_settings (expressiveness without v3 tags)
const MOOD = {
  whisper: { stability: 0.6, similarity_boost: 0.85, style: 0.15, use_speaker_boost: true },
  excited: { stability: 0.32, similarity_boost: 0.8, style: 0.7, use_speaker_boost: true },
  curious: { stability: 0.45, similarity_boost: 0.8, style: 0.5, use_speaker_boost: true },
  serious: { stability: 0.5, similarity_boost: 0.85, style: 0.35, use_speaker_boost: true },
  confident: { stability: 0.4, similarity_boost: 0.82, style: 0.55, use_speaker_boost: true },
  warm: { stability: 0.5, similarity_boost: 0.85, style: 0.45, use_speaker_boost: true },
};

// The narrator VO — ONE clear line per moment. Cast: Sandesh, Ravindra, Pushpa, Ashwini.
const LINES = [
  { id: "n01", beat: 1, mood: "excited", text: "That's Trace — and it is not a chatbot you query." },
  { id: "n02", beat: 1, mood: "excited", text: "It's an agent that watches your team, and speaks up the second you contradict yourselves." },
  { id: "n03", beat: 1, mood: "confident", text: "Observe. Remember. Reason. Detect. Guard. Learn." },
  { id: "n04", beat: 2, mood: "warm", text: "Here's a real engineering team — Sandesh, Ravindra, Pushpa and Ashwini — shipping fast across Slack and Discord." },
  { id: "n05", beat: 2, mood: "curious", text: "Nobody is querying anything. Trace is simply watching." },
  { id: "n06", beat: 3, mood: "excited", text: "Sandesh reaches for MongoDB on the new billing service. But months ago, Ravindra standardized every new service on PostgreSQL. Trace catches it instantly — cited, dated, and owned." },
  { id: "n07", beat: 4, mood: "excited", text: "Then Sandesh pushes on-prem for a big customer. But the roadmap said cloud-only. Trace flags the reversal, and loops in Ashwini." },
  { id: "n08", beat: 5, mood: "excited", text: "Pushpa starts a retry queue — one that Ravindra already shipped in platform-core. Duplicate work, caught before a day is wasted." },
  { id: "n09", beat: 6, mood: "curious", text: "And when you do want to ask, it answers — with citations." },
  { id: "n10", beat: 6, mood: "confident", text: "What did we decide about our database? It traces straight back to Ravindra's Q1 call." },
  { id: "n11", beat: 7, mood: "serious", text: "And when there is nothing to say, it says nothing." },
  { id: "n12", beat: 8, mood: "confident", text: "Underneath is one Cognee temporal knowledge graph." },
  { id: "n13", beat: 8, mood: "excited", text: "When you drift, it walks the graph back to the exact decision, and pulls it into the light." },
  { id: "n14", beat: 8, mood: "serious", text: "One graph-completion call retrieves the prior decision and judges the contradiction, in a single hop." },
  { id: "n15", beat: 9, mood: "serious", text: "The money shot. A pull request migrates billing to MongoDB. Trace comments on the pull request itself, citing Ravindra's call. Decided in Slack — enforced on GitHub." },
  { id: "n16", beat: 10, mood: "excited", text: "Here's the twist. In 2026, your code is written by a whole fleet of A-I agents that forget." },
  { id: "n17", beat: 10, mood: "curious", text: "So before Cursor writes a line, it asks Trace: has the team already decided this? Postgres. Decided by Ravindra. Mongo, rejected. It corrects itself." },
  { id: "n18", beat: 10, mood: "confident", text: "One endpoint. Every agent. The same company brain." },
  { id: "n19", beat: 11, mood: "serious", text: "And it watches the people, not just the code. Authentication is owned by Ravindra alone — and he is on leave next month." },
  { id: "n20", beat: 11, mood: "serious", text: "Ask what breaks if he is out, and every consequence is cited to a real message." },
  { id: "n21", beat: 12, mood: "confident", text: "Trace. It observes, remembers, reasons — and it acts. Nobody has to ask." },
  { id: "n22", beat: 12, mood: "warm", text: "Your team's memory — with a pulse." },
];

const clean = (t) => t.replace(/\[[^\]]*\]/g, "").replace(/\s+/g, " ").trim();

async function tts(line) {
  const res = await fetch(`https://api.elevenlabs.io/v1/text-to-speech/${NARRATOR}?output_format=mp3_44100_128`, {
    method: "POST",
    headers: { "xi-api-key": KEY, "Content-Type": "application/json", Accept: "audio/mpeg" },
    body: JSON.stringify({ text: clean(line.text), model_id: MODEL, voice_settings: MOOD[line.mood] || MOOD.confident }),
  });
  if (!res.ok) throw new Error(`${line.id}: HTTP ${res.status} ${(await res.text()).slice(0, 160)}`);
  const buf = Buffer.from(await res.arrayBuffer());
  writeFileSync(join(OUT, `${line.id}.mp3`), buf);
  return buf.length;
}

console.log(`Generating ${LINES.length} narrator lines → ${OUT} (model=${MODEL}, voice=${NARRATOR})`);
const manifest = [];
for (const line of LINES) {
  const p = join(OUT, `${line.id}.mp3`);
  if (existsSync(p) && statSync(p).size > 2000 && !process.env.FORCE) {
    console.log(`  ${line.id} · cached`);
    manifest.push({ id: line.id, beat: line.beat, mood: line.mood, text: clean(line.text) });
    continue;
  }
  try {
    const bytes = await tts(line);
    console.log(`  ${line.id} · ${(bytes / 1024).toFixed(0)}KB · "${clean(line.text).slice(0, 48)}…"`);
    manifest.push({ id: line.id, beat: line.beat, mood: line.mood, text: clean(line.text) });
  } catch (e) {
    console.error(`  ${line.id} · FAILED: ${e.message}`);
  }
  await new Promise((r) => setTimeout(r, 400)); // gentle on rate limits
}
writeFileSync(join(OUT, "manifest.json"), JSON.stringify(manifest, null, 2));
console.log(`Done. ${manifest.length}/${LINES.length} lines. Manifest written.`);
